### PR TITLE
feat(ui5-combobox): add full keyboard handling

### PR DESCRIPTION
**Pull request by stbodurov**
The keyboard specification for the ComboBox is now fully implemented, with the addition of the following keys:

- `F4`, `ALT`+`UP`/`DOWN`- Toggles the picker.
- `PAGEDOWN` - Moves selection down by page size (10 items by default).
- `PAGEUP` - Moves selection up by page size (10 items by default).
- `HOME` - If focus is in the ComboBox, moves cursor at the beginning of text. If focus is in the picker, selects the first item.
- `END` - If focus is in the ComboBox, moves cursor at the end of text. If focus is in the picker, selects the last item.

### CODE REVIEW

**Review by d3xter666 — APPROVED** [review]

File: packages/main/src/ComboBox.js [hunk 1]
```diff
@@ -591,33 +604,37 @@ class ComboBox extends UI5Element {
 	}
 
 	async handleArrowKeyPress(event) {
```
Comment (fifoosid) [inline packages/main/src/ComboBox.js:606]: This method doesn't need to be async

File: packages/main/src/ComboBox.js [hunk 2]
```diff
@@ -591,33 +604,37 @@ class ComboBox extends UI5Element {
...
-		if (this.readonly || !this._filteredItems.length) {
+		this.handleNavKeyPress(event);
+	}
+
+	async handleNavKeyPress(event) {
```
Comment (fifoosid) [inline packages/main/src/ComboBox.js:610]: same here

**Review by fifoosid — CHANGES_REQUESTED** [review]

File: packages/main/src/ComboBox.js [hunk 3]
```diff
@@ -336,12 +340,21 @@ const metadata = {
...
  * The <code>ui5-combobox</code> provides advanced keyboard handling.
- *
- * <h4>Picker</h4>
- * If the <code>ui5-combobox</code> is focused,
- * you can open or close the drop-down by pressing <code>F4</code>, <code>ALT+UP</code> or <code>ALT+DOWN</code> keys.
  * <br>
  *
+ * <ul>
+ * <li>[F4], [ALT]+[UP], or [ALT]+[DOWN] - Opens the drop-down, if available.</li>
+ * <li>[ESC] - Closes the drop-down, if open. If closed or not enabled, cancels changes and reverts to the value which the ComboBox field had when it got the focus.</li>
```
Comment (simonarangelova) [inline packages/main/src/ComboBox.js:347]:
```suggestion
 * <li>[ESC] - Closes the dropdown, if open. If closed or not enabled, cancels changes and reverts to the value which the ComboBox field had when it got the focus.</li>
```

File: packages/main/src/ComboBox.js [hunk 4]
```diff
@@ -336,12 +340,21 @@ const metadata = {
...
+ * <li>[ENTER] or [RETURN] - If drop-down is open takes over the current matching item and closes it.</li>
```
Comment (simonarangelova) [inline packages/main/src/ComboBox.js:348]:
```suggestion
 * <li>[ENTER] or [RETURN] - If dropdown is open takes over the current matching item and closes it.</li>
```

File: packages/main/src/ComboBox.js [hunk 5]
```diff
@@ -336,12 +340,21 @@ const metadata = {
...
+ * <li>[DOWN] - Selects the next matching item in the drop-down.</li>
```
Comment (simonarangelova) [inline packages/main/src/ComboBox.js:349]:
```suggestion
 * <li>[DOWN] - Selects the next matching item in the dropdown.</li>
```

File: packages/main/src/ComboBox.js [hunk 6]
```diff
@@ -336,12 +340,21 @@ const metadata = {
...
+ * <li>[UP] - Selects the previous matching item in the drop-down.</li>
```
Comment (simonarangelova) [inline packages/main/src/ComboBox.js:350]:
```suggestion
 * <li>[UP] - Selects the previous matching item in the dropdown.</li>
```

File: packages/main/src/ComboBox.js [hunk 7]
```diff
@@ -336,12 +340,21 @@ const metadata = {
...
+ * <li>[HOME] - If focus is in the ComboBox, moves caret before the first character. If focus is in the list, highlights the first item and updates the ComboBox accordingly.</li>
+ * <li>[END] - If focus is in the ComboBox, moves caret after the last character. If focus is in the list, highlights the last item and updates the ComboBox accordingly.</li>
+ * <li>[PAGEUP] - Moves highlight up by page size (10 items by default). If focus is in the ComboBox and the list is open, does nothing.</li>
```
Comment (simonarangelova) [inline packages/main/src/ComboBox.js:353]:
```suggestion
 * <li>[PAGEUP] - Moves highlight up by page size (10 items by default). </li>
```

File: packages/main/src/ComboBox.js [hunk 8]
```diff
@@ -336,12 +340,21 @@ const metadata = {
...
+ * <li>[PAGEDOWN] - Moves highlight down by page size (10 items by default). If focus is in the ComboBox and the list is open, does nothing.</li>
```
Comment (simonarangelova) [inline packages/main/src/ComboBox.js:354]:
```suggestion
 * <li>[PAGEDOWN] - Moves highlight down by page size (10 items by default).</li>
```

**Review by simonarangelova — CHANGES_REQUESTED** [review]

File: packages/main/test/specs/ComboBox.spec.js [hunk 9]
```diff
@@ -759,4 +759,64 @@ describe("Keyboard navigation", async () => {
...
+		await input.keys(['\uE009', 'a']); // Ctrl + A
+		await input.keys('Backspace');
+
+		await input.keys("Home");
+		inputValue = await input.getProperty("value");
+		assert.strictEqual("Algeria", inputValue, "The first item should be selected on HOME");
+		
+		// Clearing typed in value to prevent default behavior of END
+		await input.keys(['\uE009', 'a']); // Ctrl + A
+		await input.keys('Backspace');
```
Comment (ndeshev) [inline packages/main/test/specs/ComboBox.spec.js:808]: you can just set the value to ""

**Comment by ndeshev** [comment]
[ALT]+[UP]/[DOWN],[F4] issues with the focus on open and text selection on close remain

File: packages/main/test/specs/ComboBox.spec.js (hunk 9 above)
Comment (stbodurov) [reply packages/main/test/specs/ComboBox.spec.js:808]: Тried https://webdriver.io/docs/api/element/setValue/, does not seem to work.

Comment (ndeshev) [reply packages/main/test/specs/ComboBox.spec.js:808]: You should set the component value property - input..setProperty("value", "");

File: packages/main/src/ComboBox.js [hunk 10]
```diff
@@ -622,20 +617,17 @@ class ComboBox extends UI5Element {
...
 
 		this._isKeyNavigation = true;
 
-		isDown(event) && this._handleArrowDown(event, indexOfItem);
-		isUp(event) && this._handleArrowUp(event, indexOfItem);
-		isPageUp(event) && this._handlePageUp(event, indexOfItem);
-		isPageDown(event) && this._handlePageDown(event, indexOfItem);
-		isHome(event) && this._handleHome(event, indexOfItem);
-		isEnd(event) && this._handleEnd(event, indexOfItem);
+		this[`_handle${event.key}`](event, indexOfItem);
```
Comment (ndeshev) [inline packages/main/src/ComboBox.js:630]: well done :)

**Review by ndeshev — COMMENTED** [review]
When F4 or Alt+Arr is used to open the picker the focus should go to the first item (and it should also be selected), it works only if there is not a value state present. If it is - the value state is focused, this is wrong.

Otherwise the changes looks fine.

File: packages/main/test/specs/ComboBox.spec.js (hunk 9 above)
Comment (ndeshev) [reply packages/main/test/specs/ComboBox.spec.js:808]: Why is this comment marked as resolved.

I was looking at the tests and want to suggest 3 things:
1. If you insist of using kbd interaction for clearing the value you can just achieve it with the backspace key as it is selected by the component functionality after the nav key is pressed on closed picker.
2.  You can get rid off the inputValue variable and just use await in the assertion
3. In general the first assertion argument is the tested thing, in this case the value property, then the second is the expected value - it is so in the first part of this test and the opposite in the second half.

Comment (stbodurov) [reply packages/main/test/specs/ComboBox.spec.js]: Implemented all suggestions in the [last commit](https://github.com/SAP/ui5-webcomponents/pull/4494/commits/683153f5315e93c7b481adbc0c591c4780dba80c).

**Review by ndeshev — APPROVED** [review]

**Pull request merged**